### PR TITLE
Etag support.

### PR DESCRIPTION
**Pull request by gowri-sundaram**

### CODE REVIEW

File: cmd/disk-cache/core/part_consumer.go [hunk 1]
```diff
@@ -31,11 +34,20 @@ func (m *partitionConsumer) Meta() *dstk.Partition {
...
+	}
+	return document, err
 }
 
 func (m *partitionConsumer) put(req *dstk.DcPutReq) (interface{}, error) {
-	return nil, m.pc.Put(req.GetKey(), req.GetValue(), req.GetTtlSeconds())
+	document := &dstk.DcDocument{
+		Value: req.GetValue(),
+		Etag: req.GetEtag(),
+		LastUpdatedEpochSeconds: time.Now().Unix(),
```
Comment (anujga) [inline cmd/disk-cache/core/part_consumer.go]: dont use `time.Now()`. this makes the system hard to reproduce and test. @harshavardhangelivi has implemented some wrapper class with ideas from https://quii.gitbook.io/learn-go-with-tests/go-fundamentals/mocking

File: pkg/bdb/wrapper.go [hunk 2]
```diff
@@ -22,18 +24,40 @@ func (w *Wrapper) Get(key []byte) ([]byte, error) {
...
+		newDocument := err == badger.ErrKeyNotFound
+		if err != nil && !newDocument {
+			return err
+		}
+		if !newDocument && (document.GetEtag() != currentDocument.GetEtag()) {
+			return badger.ErrConflict
+		}
+
+		// Set a new etag.
+		randomEtag, err := uuid.NewRandom()
```
Comment (anujga) [inline pkg/bdb/wrapper.go]: During last discussion i did insist on using random, but random string wont work. We need some notion of ordering for verification of transactions. Can you use some monotonic sequence instead. Again with ability to mock for verification

File: api/protobuf-spec/dc.proto [hunk 3]
```diff
@@ -21,6 +21,13 @@ message DcPutReq {
   bytes key = 1;
   bytes value = 2;
   float ttl_seconds = 3;
+  string etag = 4;
+}
+
+message DcDocument {
+  bytes value = 1;
+  string etag = 2;
```
Comment (anujga) [inline api/protobuf-spec/dc.proto]: can u put these 2 values in a new object called `RecordMeta`. There will be more fields like ACL and many reasons to drop the meta before propagating the message

**Review by anujga — APPROVED** [review]

File: cmd/disk-cache/core/part_consumer.go (hunk 1 above)
Comment (gowri-sundaram) [reply cmd/disk-cache/core/part_consumer.go]: Done.

File: api/protobuf-spec/dc.proto (hunk 3 above)
Comment (gowri-sundaram) [reply api/protobuf-spec/dc.proto]: Done.

File: pkg/bdb/wrapper.go (hunk 2 above)
Comment (gowri-sundaram) [reply pkg/bdb/wrapper.go]: Done.

**Pull request merged**